src/coord/src/coord.rs

```diff
@@ -332,6 +334,49 @@ fn concretize_replica_config(
     Ok(config)
 }
 
+/// Holds tables needing advancement.
+struct AdvanceTables<T> {
+    map: HashMap<GlobalId, T>,
+}
+
+impl<T: CoordTimestamp> AdvanceTables<T> {
+    fn new() -> Self {
+        Self {
+            map: HashMap::new(),
+        }
+    }
+
+    // Insert an id that should be advanced to ts. If an existing and earlier entry
+    // exists, update it to ts.
+    fn insert(&mut self, id: GlobalId, ts: T) {
+        match self.map.entry(id) {
+            Entry::Occupied(mut entry) => {
+                if entry.get().less_than(&ts) {
+                    entry.insert(ts);
+                }
+            }
+            Entry::Vacant(entry) => {
+                entry.insert(ts);
+            }
+        }
+    }
+
+    // Returns the next table to advance. Blocks forever if there are none.
+    async fn recv(&mut self) -> (GlobalId, T) {
+        // TODO(mjibson): In order to prevent starvation, we need to make this random
```

Comment:
It's more complicated, but could we keep both a map and a queue in `AdvanceTables`? We would use the queue to ensure we work off advancement requests fairly, and we use the map to keep track of what the latest timestamp is that we need to advance to.

Reply:
Yes, done.